Add tutorial for patching and overlay ports (#194)
microsoft/vcpkg-docs: vcpkg/consume/install-locally-modified-package.md

author: vicroms
ms.author: viromer
ms.date: 10/24/2023
ms.prod: vcpkg
ms.topic: tutorial
zone_pivot_group_filename: zone-pivot-groups.json
zone_pivot_groups: shell-selections

#CustomerIntent: As a vcpkg user, I want to learn how to modify a vcpkg provided dependency and install it
---

# Tutorial: Install a locally modified dependency
This tutorial guides you on modifying a vcpkg port using custom overlays.
We recommend that you read the tutorial on [packaging a
library](../get_started/get-started-packaging.md) before proceeding.

In this tutorial, you will learn to:

> [!div class="checklist"]
> * [Create an overlay port](#1---create-an-overlay-port)
> * [Get the port's source code](#2---get-the-ports-source-code)
> * [Create a temporary Git registry](#3---create-a-temporary-git-registry)
> * [Modify the necessary files](#4---modify-the-necessary-files)
> * [Generate a patch file](#5---generate-a-patch-file)
> * [Modify `portfile.cmake` to apply the patch file](#6---modify-portfilecmake-to-apply-the-patch)
> * [Install your overlay port](#7---install-your-overlay-port)

## Prerequisites

* A terminal
* vcpkg
* [Git](<https://git-scm.com/downloads>)

> * A C++ compiler
> 

Reviewer: ```suggestion
* A C++ compiler
* Completion of the [packaging tutorial](../get_started/get-started-packaging.md)

```

## 1 - Create an overlay port
The first step is to create an [overlay port](../concepts/overlay-ports.md) of
the package you want to modify.

> 1 - Create a directory to hold your overlay ports

Reviewer: ```suggestion
#### Create a directory to hold your overlay ports
```

The overlay ports directory can be created in any filesystem location of your choosing. In any step of this tutorial, replace `$OVERLAY_LOCATION` with your chosen location.

::: zone pivot="shell-powershell, shell-cmd"

> ```powershell

Reviewer: ```suggestion
```console
```

mkdir "$OVERLAY_LOCATION"
```

::: zone-end
::: zone pivot="shell-bash"

```bash
mkdir "$OVERLAY_LOCATION"
```

::: zone-end

> 2 - Copy the contents of the port into your overlay ports directory

Reviewer: ```suggestion
#### Copy the contents of the port into your overlay ports directory
```

[packaging tutorial](../get_started/get-started-packaging.md) to add dynamic
library support.

::: zone pivot="shell-powershell"

```powershell
Copy-Item -Path <path/to/vcpkg-sample-library> -Destination "$OVERLAY_LOCATION" -Recurse
```

::: zone-end
::: zone pivot="shell-cmd"

```console
xcopy <path/to/vcpkg-sample-library> "$OVERLAY_LOCATION" /E
```

::: zone-end
::: zone pivot="shell-bash"

```bash
cp -R <path/to/vcpkg-sample-library> "$OVERLAY_LOCATION"
```
::: zone-end

## 2 - Get the port's source code
To obtain the source code for the port you want to modify, run the following command:

`vcpkg install "--overlay-ports=$OVERLAY_LOCATION" vcpkg-sample-library --editable`

The output should contain a line similar to:

```Console
-- Using source at path/to/vcpkg/buildtrees/vcpkg-sample-library/src/1.0.0-b54c55c215
```

This is the location of the port's source code.

## 3 - Create a temporary Git registry
Initialize a temporary Git repository in the source code location. This is in order
to use Git to generate patch files that you can apply in the `portfile.cmake`
script. Replace `$SOURCE_PATH` with the location obtained in the previous step.

> ```bash

Reviewer: ```suggestion
```console
```

---
title: "Tutorial: Install a locally modified dependency"
description: This tutorial shows you how to modify a vcpkg port and install it using overlays"
author: vicroms
ms.author: viromer
ms.date: 10/24/2023
ms.prod: vcpkg
ms.topic: tutorial
zone_pivot_group_filename: zone-pivot-groups.json
zone_pivot_groups: shell-selections

#CustomerIntent: As a vcpkg user, I want to learn how to modify a vcpkg provided dependency and install it
---

# Tutorial: Install a locally modified dependency

> This tutorial guides you on how to modified a vcpkg port using custom overlays.

Reviewer 2: ```suggestion
This tutorial guides you on modifying a vcpkg port using custom overlays.
```

Reviewer: Also maybe add a sentence telling the reader what they will get out of following this tutorial. How can overlays help them? When should they use them?

library](../get_started/get-started-packaging.md) before proceeding.

In this tutorial, you will learn to:

> [!div class="checklist"]
> * [Create an overlay port](#1---create-an-overlay-port)
> * [Get the port's source code](#2---get-the-ports-source-code)
> * [Create a temporary Git registry](#3---create-a-temporary-git-registry)
> * [Modify the necessary files](#4---modify-the-necessary-files)
> * [Generate a patch file](#5---generate-a-patch-file)
> * [Modify `portfile.cmake` to apply the patch file](#6---modify-portfilecmake-to-apply-the-patch)
> * [Install your overlay port](#7---install-your-overlay-port)

## Prerequisites

* A terminal
* vcpkg
* [Git](<https://git-scm.com/downloads>)
* A C++ compiler

## 1 - Create an overlay port
The first step is to create an [overlay port](../concepts/overlay-ports.md) of
the package you want to modify.

1 - Create a directory to hold your overlay ports

The overlay ports directory can be created in any filesystem location of your
choosing. In any step of this tutorial replace `$OVERLAY_LOCATION` with your
chosen location.

::: zone pivot="shell-powershell, shell-cmd"

```powershell
mkdir "$OVERLAY_LOCATION"

> ```

Reviewer: This will cause the cmd zone pivot to show the PowerShell box, which is a bit weird. You should use the "console" box so it's generic.

::: zone-end
::: zone pivot="shell-bash"

```bash
mkdir "$OVERLAY_LOCATION"
```

::: zone-end

2 - Copy the contents of the port into your overlay ports directory

For this tutorial, you'll modify the `vcpkg-sample-library` port in the

> [packaging tutorial](../get_started/get-started-packaging.md) to add dynamic

Reviewer 2: Do you think this tutorial should be a prerequisite? Similarly, the get started articles should include the _previous_ article in the prerequisites section.

Reviewer: Agreed for this case, but it may not be universally true for Get started articles to require to be read in order. E.g. VS tutorials don't need to be read to do the VS Code one.

library support.

::: zone pivot="shell-powershell"

```powershell
Copy-Item -Path <path/to/vcpkg-sample-library> -Destination "$OVERLAY_LOCATION" -Recurse
```

::: zone-end
::: zone pivot="shell-cmd"

```console
xcopy <path/to/vcpkg-sample-library> "$OVERLAY_LOCATION" /E
```

::: zone-end
::: zone pivot="shell-bash"

```bash
cp -R <path/to/vcpkg-sample-library> "$OVERLAY_LOCATION"
```
::: zone-end

## 2 - Get the port's source code
To obtain the source code for the port you want to modify, run the following command:

`vcpkg install "--overlay-ports=$OVERLAY_LOCATION" vcpkg-sample-library --editable`

The output should contain a line similar to:

```Console
-- Using source at path/to/vcpkg/buildtrees/vcpkg-sample-library/src/1.0.0-b54c55c215
```

This is the location of the port's source code.

## 3 - Create a temporary Git registry

> Create a temporary Git registry in the source code location, this is in  order

Reviewer 2: ```suggestion
Initialize a temporary Git repository in the source code location. This is in  order
```

Reviewer: Also there are two spaces between "in" and "order" currently.

to use Git to generate patch files that you can apply in the `portfile.cmake`
script. Replace `$SOURCE_PATH` with the location obtained in the previous step.

```bash
cd "$SOURCE_PATH"
git init
git add .
git commit -m "Initial commit"

> ```

Reviewer: This section would apply to cmd and PowerShell too, so maybe use the "console" type rather than "bash"?

)

vcpkg_cmake_configure(
    SOURCE_PATH "${SOURCE_PATH}"
)

vcpkg_cmake_install()
vcpkg_copy_pdbs()
vcpkg_cmake_config_fixup(PACKAGE_NAME "my_sample_lib")

file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/include")
file(
    INSTALL "${SOURCE_PATH}/LICENSE" 
    DESTINATION "${CURRENT_PACKAGES_DIR}/share/${PORT}" 
    RENAME copyright
)
configure_file("${CMAKE_CURRENT_LIST_DIR}/usage" "${CURRENT_PACKAGES_DIR}/share/${PORT}/usage" COPYONLY)

```

## 7 - Install your overlay port
Run the following command:

`vcpkg install "--overlay-ports=$OVERLAY_LOCATION" vcpkg-sample-library`

Make sure to remove the `--editable` flag, so that the patch file applies to a
clean copy of the source code. You should see a line about applying your patch
file to the source code in the output:

```Console
-- Cleaning sources at D:/Work/vcpkg/buildtrees/vcpkg-sample-library/src/1.0.0-8f646312ed.clean. Use --editable to skip cleaning for the packages you specify.
-- Extracting source D:/Work/vcpkg/downloads/Microsoft-vcpkg-docs-1.0.0.tar.gz
-- Applying patch add-dynamic-lib-support.patch
```

## Next steps

And that's it! You have installed a locally modified dependency using a patch
file.

> Here are some tasks to try next:;

Reviewer 2: ```suggestion
Here are some tasks to try next:
```
